### DISCUSSION TITLE
E3Sm Diags global_time_series issue

### DISCUSSION THREAD

**Discussion by keziming** [E3SM-Project/zppy#391] [Questions] [answered]
Could anyone help me with the issue running zppy? I see Luke Van Roekel post this 6 months ago, but no solution available. When I run the E3SM Diags on Chrysalis, all the sections runs well except the last part "global_time_series". Thank you in advance! You can find the error information at /lcrc/group/e3sm/ac.zke/E3SM_simulations/01052023.pi-control.MAM5-SICE.ke.maint-2.0-mam5-SICE/post/scripts/global_time_series_0501-0700.o271640

Scanning files ...
mpaso.glb.050101-051012.nc
Setting reference time units to days since 0001-01-01 00:00:00
Traceback (most recent call last):
  File "/lcrc/soft/climate/e3sm-unified/base/envs/e3sm_unified_1.8.0_chrysalis/bin/cdscan", line 1843, in <module>
    main(sys.argv)
  File "/lcrc/soft/climate/e3sm-unified/base/envs/e3sm_unified_1.8.0_chrysalis/bin/cdscan", line 1309, in main
    vartime[0],
  File "/lcrc/soft/climate/e3sm-unified/base/envs/e3sm_unified_1.8.0_chrysalis/lib/python3.10/site-packages/cdms2/axis.py", line 2212, in __getitem__
    raise IndexError('index out of bounds')
IndexError: index out of bounds

**Comment by forsyth2** [comment, marked as answer]
@keziming What is the path to your configuration file? If I have that, I can see if I can reproduce the problem.

**Reply by keziming** [reply]
@forsyth2 the file is under /lcrc/group/e3sm/ac.zke/E3SM_simulations/01052023.pi-control.MAM5-SICE.ke.maint-2.0-mam5-SICE/

**Reply by forsyth2** [reply]
I was able to reproduce the error, but I'm having difficult figuring out why it's happening. The `IndexError` is not particularly informative.

Here is a copy of my `cfg`:
```
[default]
input = /lcrc/group/e3sm/ac.zke/E3SM_simulations/01052023.pi-control.MAM5-SICE.ke.maint-2.0-mam5-SICE
input_subdir = run
#output = /lcrc/group/e3sm/ac.zke/E3SM_simulations/01052023.pi-control.MAM5-SICE.ke.maint-2.0-mam5-SICE
output = /lcrc/group/e3sm/ac.forsyth2/zppy_support/01052023.pi-control.MAM5-SICE.ke.maint-2.0-mam5-SICE
case = 01052023.pi-control.MAM5-SICE.ke.maint-2.0-mam5-SICE
#www = /lcrc/group/e3sm/public_html/diagnostic_output/ac.zke/
www = /lcrc/group/e3sm/public_html/diagnostic_output/ac.forsyth2/zppy_support/
partition = compute
#environment_commands = "source /lcrc/soft/climate/e3sm-unified/base/etc/profile.d/conda.sh; conda activate e3sm_unified_1.6.0_nompi"


[ts]
active = True
years = "0501:0700:10",
mapping_file = /home/ac.zender/data/maps/map_ne30pg2_to_cmip6_180x360_aave.20200201.nc

  [[ atm_monthly_glb ]]
  input_subdir = "run"
  input_files = "eam.h0"
  mapping_file = "glb"
  vars = "FSNTOA,FLUT,FSNT,FLNT,FSNS,FLNS,SHFLX,QFLX,TAUX,TAUY,PRECC,PRECL,PRECSC,PRECSL,TS,TREFHT,CLDTOT,CLDHGH,CLDMED,CLDLOW"

[mpas_analysis]
#active = False
active = True
walltime = "24:00:00"
parallelTaskCount = 6
anomalyRefYear = 0501
ts_years = "0501-0550","0501-0600","0501-0650","0501-0700",
enso_years = "0501-0550","0501-0600","0501-0650","0501-0700",
climo_years = "0501-0550","0501-0600","0501-0650","0501-0700",
mesh = "EC30to60E2r2"
partition = compute

[global_time_series]
active = True
ts_num_years = 10
figstr= "M5SICE.pi.0001.0200"
experiment_name = "M5.SICE.pi"
years = "0501-0700",
ts_years = "0501-0550","0501-0600","0501-0650","0501-0700",
climo_years = "0501-0550","0501-0600","0501-0650","0501-0700",
moc_file = "mocTimeSeries_0501-0700.nc"
```

When @vanroekel ran into this issue, it turned out to be because some years didn't match up -- i.e., #366. That doesn't seem to be the case here. Your years appear to match up between `mpas_analysis` and `global_time_series`. (Typically, we have `climo_years` not go all the way back to the start year, but I don't see why doing so would pose a problem.)

The status file indicates the error as being in `/lcrc/group/e3sm/ac.forsyth2/zppy_support/01052023.pi-control.MAM5-SICE.ke.maint-2.0-mam5-SICE/post/scripts/global_time_series_0501-0700.bash` at:

```
    echo 'Create xml for for ocn'
    export CDMS_NO_MPI=true
    cd ${case_dir}/post/ocn/glb/ts/monthly/${ts_num_years}yr
    cdscan -x glb.xml mpaso.glb*.nc
    if [ $? != 0 ]; then
      cd /lcrc/group/e3sm/ac.forsyth2/zppy_support/01052023.pi-control.MAM5-SICE.ke.maint-2.0-mam5-SICE/post/scripts
      echo 'ERROR (3)' > global_time_series_0501-0700.status
      exit 3
    fi
```
So, it's failing on that `cdscan -x glb.xml mpaso.glb*.nc` line.

```
$ ls /lcrc/group/e3sm/ac.forsyth2/zppy_support/01052023.pi-control.MAM5-SICE.ke.maint-2.0-mam5-SICE/post/ocn/glb/ts/monthly/10yr
mpaso.glb.050101-051012.nc  mpaso.glb.054101-055012.nc  mpaso.glb.058101-059012.nc  mpaso.glb.062101-063012.nc  mpaso.glb.066101-067012.nc
mpaso.glb.051101-052012.nc  mpaso.glb.055101-056012.nc  mpaso.glb.059101-060012.nc  mpaso.glb.063101-064012.nc  mpaso.glb.067101-068012.nc
mpaso.glb.052101-053012.nc  mpaso.glb.056101-057012.nc  mpaso.glb.060101-061012.nc  mpaso.glb.064101-065012.nc  mpaso.glb.068101-069012.nc
mpaso.glb.053101-054012.nc  mpaso.glb.057101-058012.nc  mpaso.glb.061101-062012.nc  mpaso.glb.065101-066012.nc  mpaso.glb.069101-070012.nc
```
It seems like everything is there... in any case, the error file shows it failing on the very first file, `mpaso.glb.050101-051012.nc`. The only thing I can think of that could possibly be a problem here is if each year in that file doesn't have exactly 12 months.

@chengzhuzhang @czender I'll see if I can find out any more on this, but am I missing anything obvious? Thanks!

**Reply by czender** [reply]
I don't know what `cdscan -x glb.xml mpaso.glb*.nc` is supposed to do. I suggest running on only the first file in a debugger to see exactly where (and thus why, hopefully) it fails.

**Reply by forsyth2** [reply]
> I don't know what `cdscan -x glb.xml mpaso.glb*.nc` is supposed to do.

If I recall correctly, it runs `cdscan` on all the `mpaso.glb*.nc` files before outputting to `glb.xml`.

> The only thing I can think of that could possibly be a problem here is if each year in that file doesn't have exactly 12 months.

Ah, this indeed seems to be the case:
```
$ ncdump -h mpaso.glb.050101-051012.nc
netcdf mpaso.glb.050101-051012 {
dimensions:
	time = UNLIMITED ; // (0 currently)
	nbnd = 2 ;
variables:
	double time(time) ;
		time:long_name = "time" ;
		time:units = "days since 0501-01-01 00:00:00" ;
		time:calendar = "noleap" ;
		time:bounds = "time_bnds" ;
	double time_bnds(time, nbnd) ;
		time_bnds:long_name = "time interval endpoints" ;
	double ohc(time) ;
		ohc:long_name = "total ocean heat content" ;
		ohc:units = "J" ;
	double volume(time) ;
		volume:long_name = "sum of the volumeCell variable over the full domain, used to normalize global statistics" ;
		volume:units = "m^3" ;
}

```
Notice `time = UNLIMITED ; // (0 currently)`. It looks like there's no information in the file.

**Reply by keziming** [reply]
@forsyth2 Thank you for pointing out this issue. It looks like that all files have the same issue under /lcrc/group/e3sm/ac.zke/E3SM_simulations/01052023.pi-control.MAM5-SICE.ke.maint-2.0-mam5-SICE/post/ocn/glb/ts/monthly/10yr

Is this due to some settings goes wrong in the CFG file? /lcrc/group/e3sm/ac.zke/E3SM_simulations/01052023.pi-control.MAM5-SICE.ke.maint-2.0-mam5-SICE/01052023.pi-control.MAM5-SICE.ke.maint-2.0-mam5-SICE

Thanks!

Ziming

**Reply by forsyth2** [reply]
Debugging
-----------

MPAS-Analysis ran successfully; generated plots are at https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_support/01052023.pi-control.MAM5-SICE.ke.maint-2.0-mam5-SICE/mpas_analysis/. 

The files in `/lcrc/group/e3sm/ac.forsyth2/zppy_support/01052023.pi-control.MAM5-SICE.ke.maint-2.0-mam5-SICE/post/ocn/glb/ts/monthly/10yr` are empty, however. 

Those files are generated by `ocean_month.py` (https://github.com/E3SM-Project/zppy/blob/main/zppy/templates/global_time_series.bash#L59), which uses the specified input (https://github.com/E3SM-Project/zppy/blob/main/zppy/templates/global_time_series.bash#L58). For my instantiation of the template, that input is `/lcrc/group/e3sm/ac.zke/E3SM_simulations/01052023.pi-control.MAM5-SICE.ke.maint-2.0-mam5-SICE/archive/ocn/hist`.

`/lcrc/group/e3sm/ac.zke/E3SM_simulations/01052023.pi-control.MAM5-SICE.ke.maint-2.0-mam5-SICE/` does not have an `archive` directory. 

Conclusion
-----------

Note that the step-by-step guide to running E3SM specifies to create an `archive` directory via short-term archiving (directions at https://acme-climate.atlassian.net/wiki/spaces/ED/pages/2309226536/Running+E3SM+step-by-step+guide#Short-Term-Archiving).

The alternative is to set `input_subdir` in the `cfg` manually (https://github.com/E3SM-Project/zppy/blob/main/zppy/templates/default.ini#L276)

**Reply by forsyth2** [reply]
> Note that the step-by-step guide to running E3SM specifies to create an `archive` directory via short-term archiving

Created #393 to highlight this issue, which other users have run into before.

**Reply by keziming** [reply]
@forsyth2 Thank you very much for your efforts!
In my specific case (modified maint-2.0 version), "case.st_archive" is malfunction. Therefore, I used your "alternative" setting to set "input_subdir = "run"". After this change, the mpaso.glb.*.nc files become normal (120 time in each file). However, the global time series plot still can not be done due to another error showing below. Also, you can find the error file at /lcrc/group/e3sm/ac.zke/E3SM_simulations/01052023.pi-control.MAM5-SICE.ke.maint-2.0-mam5-SICE/post/scripts/global_time_series_0501-0700.o274002

glb.xml written
Copy moc file
Update time series figures
/lcrc/group/e3sm/ac.zke/E3SM_simulations/01052023.pi-control.MAM5-SICE.ke.maint-2.0-mam5-SICE/post/atm/glb/ts/monthly/10yr/glb.xml
RESTOM
RESSURF
TREFHT
FSNTOA
FLUT
PRECC
PRECL
QFLX
M5.SICE.pi
M5.SICE.pi
M5.SICE.pi
Traceback (most recent call last):
  File "/lcrc/group/e3sm/ac.zke/E3SM_simulations/01052023.pi-control.MAM5-SICE.ke.maint-2.0-mam5-SICE/post/scripts/global_time_series_0501-0700_dir/coupled_global.py", line 405, in <module>
    ax.plot(year, var, lw=1.5, marker=None, c=exp["color"], label=exp["name"])
  File "/lcrc/soft/climate/e3sm-unified/base/envs/e3sm_unified_1.6.0_nompi/lib/python3.9/site-packages/matplotlib/axes/_axes.py", line 1632, in plot
    lines = [*self._get_lines(*args, data=data, **kwargs)]
  File "/lcrc/soft/climate/e3sm-unified/base/envs/e3sm_unified_1.6.0_nompi/lib/python3.9/site-packages/matplotlib/axes/_base.py", line 312, in __call__
    yield from self._plot_args(this, kwargs)
  File "/lcrc/soft/climate/e3sm-unified/base/envs/e3sm_unified_1.6.0_nompi/lib/python3.9/site-packages/matplotlib/axes/_base.py", line 498, in _plot_args
    raise ValueError(f"x and y must have same first dimension, but "
ValueError: x and y must have same first dimension, but have shapes (300,) and (200,)

**Reply by forsyth2** [reply]
It looks like you're using an older version of `zppy`. Using the latest version, I was able to generate the plots: https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.forsyth2/zppy_support/01052023.pi-control.MAM5-SICE.ke.maint-2.0-mam5-SICE/global_time_series/global_time_series_0501-0700_results/M5SICE.pi.0001.0200.pdf

**Reply by keziming** [reply]
@forsyth2 with latest version, it works perfectly. Thank you so much!

**Reply by forsyth2** [reply]
Glad it works!
